packages/@tinacms/toolkit/src/packages/form-builder/FormBuilder.tsx

```diff
@@ -83,6 +83,15 @@ export const FormBuilder: FC<FormBuilderProps> = ({
     [tinaForm]
   )
 
+  /**
+   * On unmount, clear the `onbeforeunload` event (if one is lingering)
+   */
+  React.useEffect(() => {
+    return () => {
+      window.onbeforeunload = undefined
```

Comment:
I wonder if it may be better to do register this like:
```window.addEventListener('beforeunload', function (e) {
```
That way, if there's other stuff happening in beforeUnload, we don't override it.

Also, Rather than unregistering/reregistering the function anytime the form state changes, should we register the form once, and check on formState within the unload function?

Reply:
I originally tried to use `addEventListener()`, but ran into issues with `addEventListener()` being added too many times and `removeEventListener()` not being called enough.  `window.onbeforeunload` ensured I only ever added and removed a single instance which worked well with how `<FormSpy />` works.

"check on formState within the unload function" - outside of the `<FormSpy />` component?

Reply:
@jamespohalloran we ended up using `finalForm.subscribe` instead of `FormSpy` for code quality reasons, but ultimately they both work the same way and only trigger when the `pristine` state changes and not every time the values change. It does still add/remove the event listeners any time that state changes, but this doesn't strike me as a terribly expensive operation